pkg/daemon/cleanup/cleanup.go

```diff
@@ -72,47 +96,55 @@ func Deconfigure(cfg *rest.Config, ctx context.Context, configMapNamespace strin
 	return nil
 }
 
-// getServicesToRemove returns a list of services that should be removed as part of the cleanup process
-// returns the merged Data of the latest ConfigMap, and the ConfigMap specified by the node's version annotation
-func getServicesToRemove(ctx context.Context, cli client.Client, node *core.Node, configMapNamespace string) ([]servicescm.Service, error) {
-	// get data from the latest services ConfigMap
-	latestCM, err := servicescm.GetLatest(cli, ctx, configMapNamespace)
+// ensureEnvVarsAreRemoved removes all WICD configured ENV variables from this instance
+func ensureEnvVarsAreRemoved(cm *core.ConfigMap) error {
+	cmData, err := servicescm.Parse(cm.Data)
 	if err != nil {
-		return nil, fmt.Errorf("cannot get latest services ConfigMap from namespace %s: %w", configMapNamespace, err)
+		return err
 	}
-	latestCMData, err := servicescm.Parse(latestCM.Data)
+	if len(cmData.EnvironmentVars) == 0 {
+		// no-op
+		return nil
+	}
+	registryKey, err := registry.OpenKey(registry.LOCAL_MACHINE, controller.SystemEnvVarRegistryPath, registry.ALL_ACCESS)
 	if err != nil {
-		return nil, err
+		return fmt.Errorf("unable to open Windows system registry key %s: %w",
+			controller.SystemEnvVarRegistryPath, err)
 	}
-
-	// attempt to get the ConfigMap specified by the version annotation
-	var versionCM core.ConfigMap
-	var versionCMData *servicescm.Data
-	err = func() error {
-		if node == nil {
-			return fmt.Errorf("no node object present")
-		}
-		version, present := node.Annotations[metadata.VersionAnnotation]
-		if !present {
-			return fmt.Errorf("node is missing version annotation")
+	defer func() { // always close the registry key, without swallowing any error returned before the defer call
+		closeErr := registryKey.Close()
+		if closeErr != nil {
+			klog.Errorf("could not close key %v: %v", registryKey, closeErr)
 		}
-		err = cli.Get(ctx, client.ObjectKey{Namespace: configMapNamespace, Name: servicescm.NamePrefix + version},
-			&versionCM)
-		if err != nil {
-			return err
-		}
-		versionCMData, err = servicescm.Parse(versionCM.Data)
+	}()
+	restartRequired, err := controller.EnsureEnvVarsAreRemoved(registryKey, cmData.SupportedEnvVars)
+	if err != nil {
+		return err
+	}
+	// rebooting instance to set the environment variables at the process level as expected
+	if restartRequired {
+		restartCmd := "Restart-Computer -Force"
```

Comment:
Should WICD be doing this? Doesn't it signal to WMCO that a restart is necessary in other places?

Reply:
Sigh, this is why I found this odd. WICD cannot be restarting instances otherwise we could have multiple or all nodes restarting at the same time.

Reply:
AFAIK WICD will not be restating multiple instances through this code path as we only cleanup one instance at a time.

Reply:
We shouldn't be spreading reboot responsibility in this manner.

Reply:
Or WMCO directly does the restart when it invokes the cleanup command.

Reply:
The last step of cleanup is deleting node object itself.

Reply:
In case there is no node to annotate

Reply:
Are you suggesting WMCO always restart the instance after cleanup is executed and if proxy was enabled?

Reply:
WMCO is what deletes the node object, so shouldn't it be present? If it is not present, can't we skip rebooting?